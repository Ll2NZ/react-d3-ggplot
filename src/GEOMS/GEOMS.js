import React            from "react";
import PropTypes        from "prop-types";
import { ChartContext } from "../_context/";

function GEOMS(props) {
        return(
            <ChartContext.Provider value={ props }>
                <svg width={ props.dimensions.width } height={ props.dimensions.height }>
                    { props.children }
                </svg>
            </ChartContext.Provider>
        );
}

GEOMS.propTypes = {
    dimensions: PropTypes.object,
    children: PropTypes.any
};

export default GEOMS;
