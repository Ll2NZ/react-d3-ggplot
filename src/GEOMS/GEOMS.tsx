import React, { ReactChildren, ReactComponentElement, ReactChild } from "react";
import { ChartContext, IContext } from "../_context/";

interface IProps extends IContext {
  children?: ReactChild[] | ReactChild;
}

function GEOMS(props: IProps) {
  if (!props.data || !props.aes || !props.dimensions) {
    throw new Error(`GEOMS must have the following properties: data, aes, and dimension`);
  }

  if (!props.children) {
    throw new Error("GEOMS expects at least one child components such as: <Rects/> or <Points/>");
  }

  return (
    <ChartContext.Provider value={props}>
      <svg width={props.dimensions.width} height={props.dimensions.height}>
        {props.children}
      </svg>
    </ChartContext.Provider>
  );
}

export default GEOMS;
