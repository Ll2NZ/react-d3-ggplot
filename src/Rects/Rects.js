import React              from "react";
import PropTypes          from "prop-types";
import { Component }      from "react";
import { select }         from "d3-selection";
import { ScalesConsumer } from "../Context/";

class Rects extends Component{
    static contextType = ScalesConsumer;

    static defaultProps = {
        className: null,
        color: "orange"
    }

    static propTypes = {
        className: PropTypes.string,
        color: PropTypes.string,
        createScaleType: PropTypes.func
    }

    appendRects(){
        // create x and y scales
        let { aes, scaleTypes } = this.context;
        let xScale              = this.props.createScaleType(aes[0], scaleTypes[0]);
        let yScale              = this.props.createScaleType(aes[1], scaleTypes[1]);

        // get other props needed for rects
        let { data, dimensions }  = this.context;
        let { color }             = this.props;

        // spreads our points across our x and y axes visually
        xScale.range([dimensions.padding, dimensions.width - dimensions.padding]);
        yScale.range([dimensions.height - dimensions.padding, dimensions.padding]);

        // clear graph for next set of data points if we have data
        if(this.node.children.length > 0)
            select(this.node).selectAll("rect").remove();

        // append new visualization
        select(this.node)
            .selectAll("rect")
            .data(data)
            .enter()
            .append("rect")
            .attr("width", xScale.bandwidth())
            .attr("height", d => yScale.range()[0] - yScale(d[aes[1]]))
            .attr("x", d => xScale(d[aes[0]]))
            .attr("y", d => yScale(d[aes[1]]))
            .attr("fill", color);
    }

    render(){
        return(
            <g
                ref={ node => this.node = node }
                className={ this.props.className }
            >
            </g>
        );
    }

    componentDidMount(){
        this.appendRects();
    }

    componentDidUpdate(){
        this.appendRects();
    }
}

export default Rects;
