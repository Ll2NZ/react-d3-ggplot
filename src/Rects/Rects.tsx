import React from "react";
import { useContext } from "react";
import { useScale } from "../_hooks/";
import { ChartContext } from "../_context/";

interface IProps {
  fill?: string;
}

function Rects(props: IProps) {
  const context = useContext(ChartContext);
  const xScale: any = useScale(context, "XAxis");
  const yScale: any = useScale(context, "YAxis");

  // will change once we start testing
  if (!xScale || !yScale) {
    return <h1>whoops</h1>;
  }

  return (
    <g>
      {context.data.map((datum, index) => (
        <rect
          key={index}
          width={xScale.bandwidth()} // this works for ordinal scales but we need to account for time and number scales as well
          height={yScale.range()[0] - yScale(datum[context.aes[1]])}
          x={xScale(datum[context.aes[0]])}
          y={yScale(datum[context.aes[1]])}
          fill={!props.fill ? "cyan" : props.fill}
        />
      ))}
    </g>
  );
}

export default Rects;
