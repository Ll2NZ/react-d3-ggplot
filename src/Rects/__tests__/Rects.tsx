// ALE complains when we just use import React from "react";
import * as React from "react";
import Rects from "../Rects";
import { GEOMS } from "../../GEOMS/";
import { render, cleanup } from "react-testing-library";

afterEach(cleanup);

// this data is only for an ordinal scale.
// Iternally, we are using xScale.bandwidth() to calculate the width of each <rect>
// But that method only exists on ordinal scales. Hence, the x-values being strings for now
// until we fix this issue
const validState = {
  data: [{ x: "developer", y: 40 }, { x: "ceo", y: 45 }],
  aes: ["x", "y"],
  dimensions: { width: 500, height: 300, padding: 50 }
};

test("<Rects/> renders a <g> and <rect>", () => {
  // We don't care about testing <GEOMS/> here.
  const { container } = render(
    <GEOMS {...validState}>
      <Rects />
    </GEOMS>
  );

  expect(container.firstChild).toMatchSnapshot();
});
