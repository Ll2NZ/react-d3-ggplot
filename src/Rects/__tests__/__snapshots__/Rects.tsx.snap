// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rects/> renders a <g> and <rect> 1`] = `
<svg
  height="300"
  width="500"
>
  <g>
    <rect
      fill="cyan"
      height="0"
      width="80"
      x="130"
      y="250"
    />
    <rect
      fill="cyan"
      height="200"
      width="80"
      x="290"
      y="50"
    />
  </g>
</svg>
`;
